In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [2]:
wns=pd.read_csv('wns_data.csv')
wns.head(10)

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   
5        58896          Analytics   region_2        Bachelor's      m   
6        20379         Operations  region_20        Bachelor's      f   
7        16290         Operations  region_34  Master's & above      m   
8        73202          Analytics  region_20        Bachelor's      m   
9        28911  Sales & Marketing   region_1  Master's & above      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   
5            sourcing                2   31                   3.0   
6               other                1   31                   3.0   
7            sourcing                1   33                   3.0   
8               other                1   28                   4.0   
9            sourcing                1   32                   5.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   
5                  7              0            0                  85   
6                  5              0            0                  59   
7                  6              0            0                  63   
8                  5              0            0                  83   
9                  5              1            0                  54   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0  
5            0  
6            0  
7            0  
8            0  
9            0

In [3]:
(((wns.isna().sum())/len(wns))*100).round(3)

employee_id             0.000
department              0.000
region                  0.000
education               4.395
gender                  0.000
recruitment_channel     0.000
no_of_trainings         0.000
age                     0.000
previous_year_rating    7.524
length_of_service       0.000
KPIs_met >80%           0.000
awards_won?             0.000
avg_training_score      0.000
is_promoted             0.000
dtype: float64

In [4]:
wns.describe()

employee_id  no_of_trainings           age  previous_year_rating  \
count  54808.000000     54808.000000  54808.000000          50684.000000   
mean   39195.830627         1.253011     34.803915              3.329256   
std    22586.581449         0.609264      7.660169              1.259993   
min        1.000000         1.000000     20.000000              1.000000   
25%    19669.750000         1.000000     29.000000              3.000000   
50%    39225.500000         1.000000     33.000000              3.000000   
75%    58730.500000         1.000000     39.000000              4.000000   
max    78298.000000        10.000000     60.000000              5.000000   

       length_of_service  KPIs_met >80%   awards_won?  avg_training_score  \
count       54808.000000   54808.000000  54808.000000        54808.000000   
mean            5.865512       0.351974      0.023172           63.386750   
std             4.265094       0.477590      0.150450           13.371559   
min             1.000000       0.000000      0.000000           39.000000   
25%             3.000000       0.000000      0.000000           51.000000   
50%             5.000000       0.000000      0.000000           60.000000   
75%             7.000000       1.000000      0.000000           76.000000   
max            37.000000       1.000000      1.000000           99.000000   

        is_promoted  
count  54808.000000  
mean       0.085170  
std        0.279137  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [5]:
wns['previous_year_rating'].median()

3.0

In [6]:
wns['education'].mode()[0]

"Bachelor's"

In [7]:
wns.isna().sum()

employee_id                0
department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64

In [8]:
wns.keys()

Index(['employee_id', 'department', 'region', 'education', 'gender',
       'recruitment_channel', 'no_of_trainings', 'age', 'previous_year_rating',
       'length_of_service', 'KPIs_met >80%', 'awards_won?',
       'avg_training_score', 'is_promoted'],
      dtype='object')

In [9]:
wns['previous_year_rating']=wns['previous_year_rating'].fillna(3.0)

In [10]:
wns['education']=wns['education'].fillna(wns['education'].mode()[0])

In [11]:
wns.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

### 1.How many employees are met the KPI?

In [12]:
wns['KPIs_met >80%'].value_counts()

0    35517
1    19291
Name: KPIs_met >80%, dtype: int64

### Ans: 19291 employees have met the kpi

## 2Q  Employees promoted in each department.

In [13]:
wns['is_promoted'].value_counts()

0    50140
1     4668
Name: is_promoted, dtype: int64

In [14]:
wns['department'].value_counts()

Sales & Marketing    16840
Operations           11348
Procurement           7138
Technology            7138
Analytics             5352
Finance               2536
HR                    2418
Legal                 1039
R&D                    999
Name: department, dtype: int64

In [15]:
freq=pd.crosstab(index=wns['is_promoted'],columns=wns['department'])
freq

department   Analytics  Finance    HR  Legal  Operations  Procurement  R&D  \
is_promoted                                                                  
0                 4840     2330  2282    986       10325         6450  930   
1                  512      206   136     53        1023          688   69   

department   Sales & Marketing  Technology  
is_promoted                                 
0                        15627        6370  
1                         1213         768

## 3Q . Average age of employee working in each department

In [16]:
wns.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [17]:
wns.keys()

Index(['employee_id', 'department', 'region', 'education', 'gender',
       'recruitment_channel', 'no_of_trainings', 'age', 'previous_year_rating',
       'length_of_service', 'KPIs_met >80%', 'awards_won?',
       'avg_training_score', 'is_promoted'],
      dtype='object')

In [18]:
wns.groupby('department')['age'].mean().reset_index()

department        age
0          Analytics  32.431241
1            Finance  32.058754
2                 HR  34.385856
3              Legal  33.832531
4         Operations  36.073669
5        Procurement  35.899972
6                R&D  32.842843
7  Sales & Marketing  34.860629
8         Technology  34.867190

## 4Q Which department have more number of people with previous year rating more than 4.0?

In [19]:
rating=wns[['department','previous_year_rating']]
rating=rating[rating['previous_year_rating']>4.0]
rating['department'].value_counts()

Operations           3198
Sales & Marketing    3173
Analytics            1420
Procurement          1375
Technology            963
HR                    568
Finance               562
R&D                   266
Legal                 216
Name: department, dtype: int64

### ANS: Operations(3198) departments have more number of employees whose previous year rating is greater than 4

## 5Q Most preferable channel for recruitment. 

In [20]:
wns['recruitment_channel'].value_counts()

other       30446
sourcing    23220
referred     1142
Name: recruitment_channel, dtype: int64

### Ans: other is the most preferrable channel for recruitment

## 6Q . Male and Female recruitment based on education qualification

In [21]:
freq=pd.crosstab(index=wns['gender'],columns=wns['education'])
freq

education  Bachelor's  Below Secondary  Master's & above
gender                                                  
f               11245              289              4778
m               27833              516             10147

## 7Q Average length of service based on region

In [22]:
avg_length=wns.groupby('region')['length_of_service'].mean()
avg_length=avg_length.reset_index()
avg_length

region  length_of_service
0    region_1           4.783607
1   region_10           5.680556
2   region_11           5.276046
3   region_12           4.176000
4   region_13           6.192221
5   region_14           5.932285
6   region_15           6.231125
7   region_16           5.690102
8   region_17           6.265075
9   region_18           4.129032
10  region_19           6.450801
11   region_2           7.314510
12  region_20           4.969412
13  region_21           5.192214
14  region_22           4.876789
15  region_23           5.876596
16  region_24           4.655512
17  region_25           4.842491
18  region_26           5.232743
19  region_27           5.018083
20  region_28           5.035660
21  region_29           4.766600
22   region_3           5.384393
23  region_30           4.952816
24  region_31           5.114729
25  region_32           4.411640
26  region_33           6.877323
27  region_34           4.349315
28   region_4           6.536700
29   region_5           5.216710
30   region_6           4.315942
31   region_7           6.226512
32   region_8           4.601527
33   region_9           3.890476

## 8Q Age wise length of service

In [23]:
wns.groupby('age')['length_of_service'].mean().reset_index()

age  length_of_service
0    20           1.513274
1    21           1.500000
2    22           1.480519
3    23           1.420561
4    24           1.618935
5    25           2.010008
6    26           2.403398
7    27           2.676689
8    28           2.969813
9    29           3.500147
10   30           4.154980
11   31           4.601302
12   32           5.269949
13   33           5.699377
14   34           5.974317
15   35           6.312431
16   36           6.316647
17   37           6.603695
18   38           6.836193
19   39           7.225369
20   40           7.236921
21   41           7.778898
22   42           8.388164
23   43           8.884073
24   44           9.318772
25   45           9.872368
26   46          10.278336
27   47          10.608618
28   48          11.323160
29   49          11.213152
30   50          11.712092
31   51          11.776350
32   52          11.866097
33   53          11.862637
34   54          12.990415
35   55          11.986395
36   56          13.670455
37   57          13.962185
38   58          14.248826
39   59          14.090909
40   60          14.096774

## 9Q Employees whose length of service is more than 10 and they have won awards

In [24]:
wns.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [25]:
wns[wns['length_of_service']>10][wns['awards_won?']==1]['employee_id'].reset_index()

index  employee_id
0    1255        45312
1    1786        62355
2    2128        21304
3    2610         3934
4    2873        70153
5    4662        19837
6    5589        62878
7    5784          381
8    5851        48831
9    6257        69987
10  10086        37232
11  10384        25576
12  11439        43699
13  13612        19999
14  16622        45370
15  16673        32222
16  17540        10335
17  18310        31792
18  19499         5827
19  20673        53052
20  21552        23774
21  21610        61540
22  23346        15413
23  23976        42640
24  25362        76122
25  25569        57761
26  25657        49276
27  25993        18179
28  26127        63286
29  27164         6647
30  28651        38118
31  29400        59411
32  29438        18334
33  30624        50422
34  30833        19374
35  31288         4031
36  31421        38312
37  32690        58129
38  33234        64357
39  33757        56022
40  34069         1724
41  34957         5878
42  36917        78133
43  37253        71273
44  37578        30197
45  38214        56712
46  43054        49851
47  44180        64453
48  45409        61796
49  46168          442
50  46612        72596
51  48463        39669
52  48606        76836
53  50326        36591
54  51937        62958
55  52128        35105

## 10Q  Promoted employee’s average training score.

In [26]:
wns.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [31]:
promoted.head()

employee_id         department     region         education gender  \
11        49017  Sales & Marketing   region_7        Bachelor's      f   
39        58304  Sales & Marketing  region_28        Bachelor's      m   
60        17673  Sales & Marketing   region_4  Master's & above      m   
66        77981            Finance  region_22        Bachelor's      m   
67        16502  Sales & Marketing  region_22        Bachelor's      m   

   recruitment_channel  no_of_trainings  age  previous_year_rating  \
11            sourcing                1   35                   5.0   
39            sourcing                1   33                   5.0   
60               other                1   50                   4.0   
66               other                1   27                   3.0   
67            sourcing                1   27                   3.0   

    length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
11                  3              1            0                  50   
39                  6              1            0                  51   
60                 17              1            0                  47   
66                  1              1            1                  58   
67                  1              0            0                  61   

    is_promoted  
11            1  
39            1  
60            1  
66            1  
67            1

In [30]:
promoted=wns[wns['is_promoted']==1]
promoted['avg_training_score'].mean()

71.32519280205655

### Ans: average training score for promoted employees is 71.32

In [32]:
promoted.groupby('employee_id')['avg_training_score'].mean().reset_index()

employee_id  avg_training_score
0              39                  90
1              47                  82
2              56                  48
3              57                  79
4              62                  50
5             116                  83
6             124                  87
7             125                  59
8             134                  87
9             142                  90
10            158                  69
11            170                  94
12            189                  76
13            208                  47
14            236                  77
15            253                  52
16            271                  69
17            272                  82
18            283                  82
19            287                  62
20            298                  96
21            312                  45
22            315                  62
23            316                  94
24            317                  47
25            319                  45
26            365                  51
27            399                  65
28            402                  57
29            406                  97
...           ...                 ...
4638        77998                  85
4639        78001                  68
4640        78004                  59
4641        78035                  81
4642        78039                  62
4643        78071                  91
4644        78095                  83
4645        78102                  68
4646        78106                  74
4647        78108                  96
4648        78115                  98
4649        78133                  72
4650        78139                  54
4651        78140                  47
4652        78156                  59
4653        78172                  77
4654        78177                  58
4655        78178                  73
4656        78182                  59
4657        78184                  69
4658        78186                  82
4659        78206                  55
4660        78216                  70
4661        78218                  93
4662        78226                  54
4663        78263                  44
4664        78281                  58
4665        78285                  94
4666        78288                  60
4667        78292                  57

[4668 rows x 2 columns]

In [29]:
wns.shape

(54808, 14)